xc/device.ml

```diff
@@ -1800,39 +1798,45 @@ module Dm_Common = struct
     debug
       "%s: should be running in the background (stdout -> syslog); (fd,pid) = %s"
       name (Forkhelpers.string_of_pidty pid);
-    begin
-      let finished = ref false in
-      let watch = Watch.value_to_appear ready_path |> Watch.map (fun _ -> ()) in
-      let start = Unix.gettimeofday () in
-      while Unix.gettimeofday () -. start < timeout && not !finished do
-        Xenops_task.check_cancelling task;
-        try
-          let (_: bool) = cancellable_watch cancel [ watch ] [] task ~xs ~timeout () in
-          let state = try xs.Xs.read ready_path with _ -> "" in
-          match ready_val with
-          | Some value ->
-            if state = value
-            then finished := true
-            else raise (Ioemu_failed
-                          (name, (Printf.sprintf "Daemon state not running (%s)" state)))
-          | None -> finished := true
-        with Watch.Timeout _ ->
-          begin match Forkhelpers.waitpid_nohang pid with
-            | 0, Unix.WEXITED 0 -> () (* still running => keep waiting *)
-            | _, Unix.WEXITED n ->
-              error "%s: unexpected exit with code: %d" name n;
-              raise (Ioemu_failed (name, "Daemon exited unexpectedly"))
-            | _, (Unix.WSIGNALED n | Unix.WSTOPPED n) ->
-              error "%s: unexpected signal: %d" name n;
-              raise (Ioemu_failed (name, "Daemon exited unexpectedly"))
-          end
-      done;
-      if not !finished then
-        raise (Ioemu_failed (name, "Timeout reached while starting daemon"))
-    end;
-    debug "Daemon initialised: %s" syslog_key;
+    debug "Daemon started: %s" syslog_key;
     pid
 
+  (* Waits for a daemon to signal startup by writing to a xenstore path
+   * (optionally with a given value) If this doesn't happen in the timeout then
+   * an exception is raised *)
+  let wait_path ~pid ~task ~name ~domid ~xs ~ready_path ?ready_val ~timeout
+      ~cancel _ =
+    let syslog_key = (Printf.sprintf "%s-%d" name domid) in
+    let finished = ref false in
+    let watch = Watch.value_to_appear ready_path |> Watch.map (fun _ -> ()) in
+    let start = Mtime_clock.counter () in
+    while (Mtime.Span.to_ms (Mtime_clock.count start)) < timeout && not !finished do
+      Xenops_task.check_cancelling task;
+      try
+        let (_: bool) = cancellable_watch cancel [ watch ] [] task ~xs ~timeout () in
+        let state = try xs.Xs.read ready_path with _ -> "" in
+        match ready_val with
+        | Some value ->
```

Comment:
This is a matter of taste, you could do:
```
match ready_val with
| Some value when value = state -> finished := true
| Some _ -> raise ..
| None -> finished := false
```

Reply:
Note that apart from changing to a monotonic clock, the wait_path function is split out from the original init_daemon but I didn't actually change any of the code. I'm not sure if GitHub has the ability to diff ignoring whitespace changes, but that would be useful.

Reply:
You can add ?w=false to the url. Good to look at the changes but you cannot comment in ignore-whitespace mode